parquet-common/src/main/java/org/apache/parquet/bytes/MultiBufferInputStream.java

```diff
@@ -0,0 +1,374 @@
+import java.util.NoSuchElementException;
+
+class MultiBufferInputStream extends ByteBufferInputStream {
+  private static final ByteBuffer EMPTY = ByteBuffer.allocate(0);
+
+  private final Collection<ByteBuffer> buffers;
+  private final long length;
+
+  private Iterator<ByteBuffer> iterator;
+  private ByteBuffer current = EMPTY;
+  private long position = 0;
+
+  private long mark = -1;
+  private long markLimit = 0;
+  private List<ByteBuffer> markBuffers = new ArrayList<>();
+
+  MultiBufferInputStream(Collection<ByteBuffer> buffers) {
+    this.buffers = buffers;
+
+    long totalLen = 0;
+    for (ByteBuffer buffer : buffers) {
+      totalLen += buffer.remaining();
+    }
+    this.length = totalLen;
+
+    this.iterator = buffers.iterator();
+
+    nextBuffer();
+  }
+
+  /**
+   * Returns the position in the stream.
+   */
+  public long position() {
+    return position;
+  }
+
+  @Override
+  public long skip(long n) throws IOException {
+    if (n <= 0) {
+      return 0;
+    }
+
+    if (current == null) {
+      return -1;
+    }
+
+    long bytesSkipped = 0;
+    while (bytesSkipped < n) {
+      if (current.remaining() > 0) {
+        long bytesToSkip = Math.min(n - bytesSkipped, current.remaining());
+        current.position(current.position() + (int) bytesToSkip);
+        bytesSkipped += bytesToSkip;
+        this.position += bytesToSkip;
+      } else if (!nextBuffer()) {
+        // there are no more buffers
+        return bytesSkipped > 0 ? bytesSkipped : -1;
+      }
+    }
+
+    return bytesSkipped;
+  }
+
+  @Override
+  public int read(ByteBuffer out) {
+    int len = out.remaining();
+    if (len <= 0) {
+      return 0;
+    }
+
+    if (current == null) {
+      return -1;
+    }
+
+    int bytesCopied = 0;
+    while (bytesCopied < len) {
+      if (current.remaining() > 0) {
+        int bytesToCopy;
+        ByteBuffer copyBuffer;
+        if (current.remaining() <= out.remaining()) {
+          // copy all of the current buffer
+          bytesToCopy = current.remaining();
+          copyBuffer = current;
+        } else {
+          // copy a slice of the current buffer
+          bytesToCopy = out.remaining();
+          copyBuffer = current.duplicate();
+          copyBuffer.limit(copyBuffer.position() + bytesToCopy);
+          current.position(copyBuffer.position() + bytesToCopy);
+        }
+
+        out.put(copyBuffer);
+        bytesCopied += bytesToCopy;
+        this.position += bytesToCopy;
+
+      } else if (!nextBuffer()) {
+        // there are no more buffers
+        return bytesCopied > 0 ? bytesCopied : -1;
+      }
+    }
+
+    return bytesCopied;
+  }
+
+  @Override
+  public ByteBuffer slice(int length) throws EOFException {
+    if (length <= 0) {
+      return EMPTY;
+    }
+
+    if (current == null) {
+      throw new EOFException();
+    }
+
+    ByteBuffer slice;
+    if (length > current.remaining()) {
+      // a copy is needed to return a single buffer
```

Comment:
Copies are sometimes needed in order to get buffers back as `ByteBuffer` and not have one massive allocation. There's no way to get around needing to copy in some cases where a read crosses the boundary between buffers.

However, cross-buffer reads are hardly ever the case because the buffers inside these input streams are on the order of megabytes, while slices from this method are on the order of bytes. Callers that need large slices can also use `sliceBuffers` to avoid the problem. The benchmarks that I ran show that this is the case in practice: this buffer management is no slower than the existing code that depends on a huge allocation.

Reply:
Another point: it's far more expensive to copy, which happens fairly often in the current read path because everything is based on byte arrays and not byte buffers. I think the copy here isn't significant enough to worry about in comparison.